core/blockchain.go

```diff
@@ -392,6 +396,14 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, genesis *Genesis
 		var diskRoot common.Hash
 		if bc.cacheConfig.SnapshotLimit > 0 {
 			diskRoot = rawdb.ReadSnapshotRoot(bc.db)
+			if bc.triedb.Scheme() == rawdb.PathScheme {
+				if err := bc.triedb.Recover(diskRoot); err != nil {
+					diskRoot = common.Hash{}
+					log.Warn("failed to recover pathdb by snapshot, reset chain from genesis", "error", err)
```

Comment:
Looks like a very danger operation to reset to genesis, if snapshot diskRoot not found.
Any mechanism to iterate back to find a usable layer?

Reply:
yes, the newest code add `bc.stateRecoverable` check, so delete it